File: bench/template/math/matrix_mul/matrix_mul.py
import json, time

def matmul(a, b):
    n, m, p = len(a), len(b[0]), len(b)
    return [
        [sum(a[i][k] * b[k][j] for k in range(p)) for j in range(m)]
        for i in range(n)
    ]

size = {{ .N }}
repeat = 10

a = [[i + j for j in range(size)] for i in range(size)]
b = [[i * j for j in range(size)] for i in range(size)]

start = time.perf_counter()
for _ in range(repeat):
    last = matmul(a, b)
duration = (time.perf_counter() - start) * 1000

print(json.dumps({
    "duration_ms": duration,
    "output": last[0][0],
}))

File: bench/template/math/matrix_mul/matrix_mul.ts
function matmul(a: number[][], b: number[][]): number[][] {
    const n = a.length, m = b[0].length, p = b.length;
    const result: number[][] = [];

    for (let i = 0; i < n; i++) {
        const row: number[] = [];
        for (let j = 0; j < m; j++) {
            let sum = 0;
            for (let k = 0; k < p; k++) {
                sum += a[i][k] * b[k][j];
            }
            row.push(sum);
        }
        result.push(row);
    }
    return result;
}

const size = {{ .N }};
const repeat = 10;

const a = Array.from({ length: size }, (_, i) =>
    Array.from({ length: size }, (_, j) => i + j)
);
const b = Array.from({ length: size }, (_, i) =>
    Array.from({ length: size }, (_, j) => i * j)
);

let last: number[][] = [];
const start = performance.now();
for (let i = 0; i < repeat; i++) {
    last = matmul(a, b);
}
const duration = performance.now() - start;

console.log(JSON.stringify({
    duration_ms: duration,
    output: last[0][0],
}));
